File: plugins/cache/cc-query-dev/cc-query/0.6.1/zig/src/session_loader.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const paths = @import("paths.zig");
const types = @import("types.zig");

/// Count result for sessions and agents
pub const Counts = struct {
    sessions: usize,
    agents: usize,
};

/// Count sessions and agents from a list of file paths
pub fn countSessionsAndAgents(files: []const []const u8, session_filter: ?[]const u8) Counts {
    var sessions: usize = 0;
    var agents: usize = 0;

    for (files) |file| {
        if (!std.mem.endsWith(u8, file, ".jsonl")) continue;

        // Get basename
        const basename = if (std.mem.lastIndexOf(u8, file, "/")) |idx|
            file[idx + 1 ..]
        else
            file;

        const is_subagent_path = std.mem.indexOf(u8, file, "/subagents/") != null;

        if (is_subagent_path and std.mem.startsWith(u8, basename, "agent-")) {
            // Subagent file: {sessionId}/subagents/agent-xxx.jsonl
            if (session_filter) |filter| {
                // Check if this subagent belongs to a filtered session
                const session_dir = if (std.mem.indexOf(u8, file, "/")) |idx|
                    file[0..idx]
                else
                    file;
                if (std.mem.startsWith(u8, session_dir, filter)) {
                    agents += 1;
                }
            } else {
                agents += 1;
            }
        } else if (!std.mem.startsWith(u8, basename, "agent-") and !is_subagent_path) {
            // Session file: {sessionId}.jsonl (top-level, not agent- prefixed)
            if (session_filter) |filter| {
                if (std.mem.startsWith(u8, basename, filter)) {
                    sessions += 1;
                }
            } else {
                sessions += 1;
            }
        }
    }

    return .{ .sessions = sessions, .agents = agents };
}

/// Find all JSONL files in a directory recursively
/// Returns owned list of owned strings
pub fn findJsonlFiles(allocator: Allocator, base_path: []const u8) ![][]const u8 {
    var files: std.ArrayListUnmanaged([]const u8) = .{};
    errdefer {
        for (files.items) |f| allocator.free(f);
        files.deinit(allocator);
    }

    // Handle relative paths by opening from CWD
    var dir = (if (std.fs.path.isAbsolute(base_path))
        std.fs.openDirAbsolute(base_path, .{ .iterate = true })
    else
        std.fs.cwd().openDir(base_path, .{ .iterate = true })) catch |err| {
        if (err == error.FileNotFound or err == error.NotDir) {
            return try files.toOwnedSlice(allocator);
        }
        return err;
    };
    defer dir.close();

    var walker = try dir.walk(allocator);
    defer walker.deinit();

    while (try walker.next()) |entry| {
        if (entry.kind == .file and std.mem.endsWith(u8, entry.path, ".jsonl")) {
            // entry.path is relative to base_path, just copy it
            const path_copy = try allocator.dupe(u8, entry.path);
            try files.append(allocator, path_copy);
        }
    }

    return try files.toOwnedSlice(allocator);
}

/// Get all project directories in ~/.claude/projects
pub fn getAllProjectDirs(allocator: Allocator) ![][]const u8 {
    const base = try paths.getClaudeProjectsBase(allocator);
    defer allocator.free(base);

    var dirs: std.ArrayListUnmanaged([]const u8) = .{};
    errdefer {
        for (dirs.items) |d| allocator.free(d);
        dirs.deinit(allocator);
    }

    var dir = std.fs.openDirAbsolute(base, .{ .iterate = true }) catch |err| {
        if (err == error.FileNotFound or err == error.NotDir) {
            return try dirs.toOwnedSlice(allocator);
        }
        return err;
    };
    defer dir.close();

    var iter = dir.iterate();
    while (try iter.next()) |entry| {
        if (entry.kind == .directory) {
            const full_path = try std.fs.path.join(allocator, &.{ base, entry.name });
            try dirs.append(allocator, full_path);
        }
    }

    return try dirs.toOwnedSlice(allocator);
}

/// Options for getSessionFiles
pub const GetSessionFilesOptions = struct {
    data_dir: ?[]const u8 = null,
};

/// Get session info and file pattern for querying
pub fn getSessionFiles(
    allocator: Allocator,
    claude_projects_dir: ?[]const u8,
    session_filter: ?[]const u8,
    options: GetSessionFilesOptions,
) !types.SessionInfo {
    // If dataDir is specified, use it directly
    if (options.data_dir) |data_dir| {
        const files = try findJsonlFiles(allocator, data_dir);
        defer {
            for (files) |f| allocator.free(f);
            allocator.free(files);
        }

        const counts = countSessionsAndAgents(files, session_filter);

        if (counts.sessions == 0 and counts.agents == 0) {
            // Check for any JSONL files
            var jsonl_count: usize = 0;
            for (files) |f| {
                if (std.mem.endsWith(u8, f, ".jsonl")) jsonl_count += 1;
            }

            if (jsonl_count == 0) {
                return .{
                    .session_count = 0,
                    .agent_count = 0,
                    .project_count = 0,
                    .file_pattern = .{ .single = try allocator.dupe(u8, "") },
                };
            }

            // Has JSONL files but don't match normal session naming
            const pattern = try std.fs.path.join(allocator, &.{ data_dir, "**/*.jsonl" });
            return .{
                .session_count = jsonl_count,
                .agent_count = 0,
                .project_count = 1,
                .file_pattern = .{ .single = pattern },
            };
        }

        if (session_filter) |filter| {
            var patterns: std.ArrayListUnmanaged([]const u8) = .{};
            errdefer {
                for (patterns.items) |p| allocator.free(p);
                patterns.deinit(allocator);
            }

            const session_pattern = try std.fmt.allocPrint(allocator, "{s}/{s}*.jsonl", .{ data_dir, filter });
            try patterns.append(allocator, session_pattern);

            if (counts.agents > 0) {
                const agent_pattern = try std.fmt.allocPrint(allocator, "{s}/{s}*/subagents/*.jsonl", .{ data_dir, filter });
                try patterns.append(allocator, agent_pattern);
            }

            return .{
                .session_count = counts.sessions,
                .agent_count = counts.agents,
                .project_count = 1,
                .file_pattern = .{ .multiple = try patterns.toOwnedSlice(allocator) },
            };
        } else {
            const pattern = try std.fs.path.join(allocator, &.{ data_dir, "**/*.jsonl" });
            return .{
                .session_count = counts.sessions,
                .agent_count = counts.agents,
                .project_count = 1,
                .file_pattern = .{ .single = pattern },
            };
        }
    }

    // If no specific project, use all projects
    if (claude_projects_dir == null) {
        const base = try paths.getClaudeProjectsBase(allocator);
        defer allocator.free(base);

        const project_dirs = try getAllProjectDirs(allocator);
        defer {
            for (project_dirs) |d| allocator.free(d);
            allocator.free(project_dirs);
        }

        var total_sessions: usize = 0;
        var total_agents: usize = 0;

        for (project_dirs) |dir| {
            const files = try findJsonlFiles(allocator, dir);
            defer {
                for (files) |f| allocator.free(f);
                allocator.free(files);
            }
            const counts = countSessionsAndAgents(files, session_filter);
            total_sessions += counts.sessions;
            total_agents += counts.agents;
        }

        if (total_sessions == 0) {
            return .{
                .session_count = 0,
                .agent_count = 0,
                .project_count = 0,
                .file_pattern = .{ .single = try allocator.dupe(u8, "") },
            };
        }

        if (session_filter) |filter| {
            var patterns: std.ArrayListUnmanaged([]const u8) = .{};
            errdefer {
                for (patterns.items) |p| allocator.free(p);
                patterns.deinit(allocator);
            }

            const session_pattern = try std.fmt.allocPrint(allocator, "{s}/*/{s}*.jsonl", .{ base, filter });
            try patterns.append(allocator, session_pattern);

            if (total_agents > 0) {
                const agent_pattern = try std.fmt.allocPrint(allocator, "{s}/*/{s}*/subagents/*.jsonl", .{ base, filter });
                try patterns.append(allocator, agent_pattern);
            }

            return .{
                .session_count = total_sessions,
                .agent_count = total_agents,
                .project_count = project_dirs.len,
                .file_pattern = .{ .multiple = try patterns.toOwnedSlice(allocator) },
            };
        } else {
            const pattern = try std.fmt.allocPrint(allocator, "{s}/*/**/*.jsonl", .{base});
            return .{
                .session_count = total_sessions,
                .agent_count = total_agents,
                .project_count = project_dirs.len,
                .file_pattern = .{ .single = pattern },
            };
        }
    }

    // Specific project directory
    const proj_dir = claude_projects_dir.?;
    const files = try findJsonlFiles(allocator, proj_dir);
    defer {
        for (files) |f| allocator.free(f);
        allocator.free(files);
    }

    const counts = countSessionsAndAgents(files, session_filter);

    if (counts.sessions == 0) {
        return .{
            .session_count = 0,
            .agent_count = 0,
            .project_count = 1,
            .file_pattern = .{ .single = try allocator.dupe(u8, "") },
        };
    }

    if (session_filter) |filter| {
        var patterns: std.ArrayListUnmanaged([]const u8) = .{};
        errdefer {
            for (patterns.items) |p| allocator.free(p);
            patterns.deinit(allocator);
        }

        const session_pattern = try std.fmt.allocPrint(allocator, "{s}/{s}*.jsonl", .{ proj_dir, filter });
        try patterns.append(allocator, session_pattern);

        if (counts.agents > 0) {
            const agent_pattern = try std.fmt.allocPrint(allocator, "{s}/{s}*/subagents/*.jsonl", .{ proj_dir, filter });
            try patterns.append(allocator, agent_pattern);
        }

        return .{
            .session_count = counts.sessions,
            .agent_count = counts.agents,
            .project_count = 1,
            .file_pattern = .{ .multiple = try patterns.toOwnedSlice(allocator) },
        };
    } else {
        const pattern = try std.fs.path.join(allocator, &.{ proj_dir, "**/*.jsonl" });
        return .{
            .session_count = counts.sessions,
            .agent_count = counts.agents,
            .project_count = 1,
            .file_pattern = .{ .single = pattern },
        };
    }
}

test "countSessionsAndAgents distinguishes file types" {
    const files = &[_][]const u8{
        "abc123.jsonl",
        "abc123/subagents/agent-xyz.jsonl",
        "def456.jsonl",
    };
    const counts = countSessionsAndAgents(files, null);
    try std.testing.expectEqual(@as(usize, 2), counts.sessions);
    try std.testing.expectEqual(@as(usize, 1), counts.agents);
}

test "countSessionsAndAgents with filter" {
    const files = &[_][]const u8{
        "abc123.jsonl",
        "abc123/subagents/agent-xyz.jsonl",
        "def456.jsonl",
    };
    const counts = countSessionsAndAgents(files, "abc");
    try std.testing.expectEqual(@as(usize, 1), counts.sessions);
    try std.testing.expectEqual(@as(usize, 1), counts.agents);
}

test "countSessionsAndAgents excludes non-jsonl" {
    const files = &[_][]const u8{
        "abc123.jsonl",
        "readme.md",
        "config.json",
    };
    const counts = countSessionsAndAgents(files, null);
    try std.testing.expectEqual(@as(usize, 1), counts.sessions);
    try std.testing.expectEqual(@as(usize, 0), counts.agents);
}

File: plugins/cache/cc-query-dev/cc-query/0.6.1/src/session-loader.js
import { readdir } from "node:fs/promises";
import { join } from "node:path";
import { homedir } from "node:os";

/**
 * Get the base Claude projects directory
 * @returns {string}
 */
export function getClaudeProjectsBase() {
  return join(homedir(), ".claude", "projects");
}

/**
 * Get all project directories
 * @returns {Promise<string[]>} Array of project directory paths
 */
export async function getAllProjectDirs() {
  const base = getClaudeProjectsBase();
  const entries = await readdir(base, { withFileTypes: true });
  return entries.filter((e) => e.isDirectory()).map((e) => join(base, e.name));
}

/**
 * Count sessions and agents from a list of file paths
 * @param {string[]} files - Array of relative file paths from project directory
 * @param {string} sessionFilter - Optional session ID prefix
 * @returns {{ sessions: number, agents: number }}
 */
function countSessionsAndAgents(files, sessionFilter = "") {
  const jsonlFiles = files.filter((e) => e.endsWith(".jsonl"));

  let sessions = 0;
  let agents = 0;

  for (const file of jsonlFiles) {
    const basename = file.split("/").pop() ?? file;
    const isSubagentPath = file.includes("/subagents/");

    if (isSubagentPath && basename.startsWith("agent-")) {
      // Subagent file: {sessionId}/subagents/agent-xxx.jsonl
      // Check if this subagent belongs to a filtered session
      if (sessionFilter) {
        const sessionDir = file.split("/")[0];
        if (sessionDir?.startsWith(sessionFilter)) {
          agents++;
        }
      } else {
        agents++;
      }
    } else if (!basename.startsWith("agent-") && !isSubagentPath) {
      // Session file: {sessionId}.jsonl (top-level, not agent- prefixed)
      if (!sessionFilter || basename.startsWith(sessionFilter)) {
        sessions++;
      }
    }
  }

  return { sessions, agents };
}

/**
 * Get session info and file pattern for querying
 * @param {string | null} claudeProjectsDir - Path to ~/.claude/projects/{slug}, or null for all projects
 * @param {string} [sessionFilter] - Optional session ID prefix
 * @param {{ dataDir?: string }} [options] - Additional options
 * @returns {Promise<{ sessionCount: number, agentCount: number, projectCount: number, filePattern: string | string[] }>}
 */
export async function getSessionFiles(
  claudeProjectsDir,
  sessionFilter = "",
  options = {},
) {
  const { dataDir } = options;

  // If dataDir is specified, use it directly as the JSONL source
  if (dataDir) {
    const entries = await readdir(dataDir, { recursive: true });
    const { sessions, agents } = countSessionsAndAgents(entries, sessionFilter);

    if (sessions === 0 && agents === 0) {
      // Check for any JSONL files at all
      const jsonlFiles = entries.filter((e) => e.endsWith(".jsonl"));
      if (jsonlFiles.length === 0) {
        return {
          sessionCount: 0,
          agentCount: 0,
          projectCount: 0,
          filePattern: "",
        };
      }
      // Has JSONL files but they don't match normal session naming - still use them
      return {
        sessionCount: jsonlFiles.length,
        agentCount: 0,
        projectCount: 1,
        filePattern: join(dataDir, "**/*.jsonl"),
      };
    }

    let filePattern;
    if (sessionFilter) {
      filePattern = [join(dataDir, `${sessionFilter}*.jsonl`)];
      if (agents > 0) {
        filePattern.push(join(dataDir, `${sessionFilter}*/subagents/*.jsonl`));
      }
    } else {
      filePattern = join(dataDir, "**/*.jsonl");
    }

    return {
      sessionCount: sessions,
      agentCount: agents,
      projectCount: 1,
      filePattern,
    };
  }
  // If no specific project, use all projects
  if (!claudeProjectsDir) {
    const base = getClaudeProjectsBase();
    const projectDirs = await getAllProjectDirs();

    let totalSessions = 0;
    let totalAgents = 0;

    for (const dir of projectDirs) {
      // Recursively find all jsonl files (includes */subagents/*.jsonl)
      const entries = await readdir(dir, { recursive: true });
      const counts = countSessionsAndAgents(entries, sessionFilter);
      totalSessions += counts.sessions;
      totalAgents += counts.agents;
    }

    if (totalSessions === 0) {
      return {
        sessionCount: 0,
        agentCount: 0,
        projectCount: 0,
        filePattern: "",
      };
    }

    // Use glob pattern for all projects (** for recursive matching)
    // When session filter is provided, include both the filtered session AND its subagents
    // Subagents are stored in {session_id}/subagents/*.jsonl
    // Only include subagents pattern if there are actually subagent files
    let filePattern;
    if (sessionFilter) {
      filePattern = [join(base, "*", `${sessionFilter}*.jsonl`)];
      if (totalAgents > 0) {
        filePattern.push(join(base, "*", `${sessionFilter}*/subagents/*.jsonl`));
      }
    } else {
      filePattern = join(base, "*", "**/*.jsonl");
    }

    return {
      sessionCount: totalSessions,
      agentCount: totalAgents,
      projectCount: projectDirs.length,
      filePattern,
    };
  }

  // Recursively find all jsonl files (includes */subagents/*.jsonl)
  const entries = await readdir(claudeProjectsDir, { recursive: true });
  const { sessions, agents } = countSessionsAndAgents(entries, sessionFilter);

  if (sessions === 0) {
    return { sessionCount: 0, agentCount: 0, projectCount: 1, filePattern: "" };
  }

  // Use glob pattern for matching
  // When session filter is provided, include both the filtered session AND its subagents
  // Subagents are stored in {session_id}/subagents/*.jsonl
  // Only include subagents pattern if there are actually subagent files
  let filePattern;
  if (sessionFilter) {
    filePattern = [join(claudeProjectsDir, `${sessionFilter}*.jsonl`)];
    if (agents > 0) {
      filePattern.push(
        join(claudeProjectsDir, `${sessionFilter}*/subagents/*.jsonl`),
      );
    }
  } else {
    filePattern = join(claudeProjectsDir, "**/*.jsonl");
  }

  return {
    sessionCount: sessions,
    agentCount: agents,
    projectCount: 1,
    filePattern,
  };
}
